### ISSUE TITLE
Provide compatibility when add new column in table without regenerating model

### ISSUE THREAD

**Issue by Yongke** [volatiletech/sqlboiler#122]
I am migrating a Rails application to Go which is huge and I am working step by step for the most busy part first.

The problem I met is that, the remain Rails part may change the table frequently (likely add new columns by other guys). When this happens, I must regenerate the model and recompile,redeploy. Otherwise query will fail with following error
```
models: failed to execute a one query for shops: could not find struct field name in mapping: HadesBillboardBannerTemplateID
```

Is it possible to provide an feature that not fail the query/update but just ignore the newly added columns?

For an running product, this is critical, because we don't wanna be out of service when just adding a new column in DB table.

**Comment by Yongke** [comment]
Update: we use a lot of 'select *' query style

**Comment by nullbio** [comment]
This *may* be possible from a code perspective, we'd have to look into it more. From memory this error is coming from the bind function, if it fails to find a the column coming back from the select in the struct then it fails. It would in theory be possible to have this fail gracefully, but we'd have to look at the repercussions it would have on other functionality as well. Also, an argument could be made that no error here is bad because it indicates a mismatch between your database and your models, and that is something that users should be aware of. 

@Yongke can you elaborate on your use case further, why are you adding columns at run time? From the top of my head this sounds like a bad idea and should be supplemented with some other mechanism (i.e. adding a row with metadata opposed to a new column). @aarondl What do you think of this from a database design perspective?

**Comment by Yongke** [comment]
@nullbio The application was evolving for some years, and most of the tables is not well designed. :(

**Comment by nullbio** [comment]
@Yongke I understand, but is there a particular reason why you are adding columns at run time, or is it just due to the bad design?

**Comment by aarondl** [comment]
This is an extremely difficult request. I mean the *entire point* of this kind of tool is to: Update schema, regenerate models. That's the workflow and the code's all been built around this assumption.


Consider the call:
https://golang.org/pkg/database/sql/#Rows.Scan
And the documentation fragment:
> Scan copies the columns in the current row into the values pointed at by dest. The number of values in dest must be the same as the number of columns in Rows.

This makes it difficult. If we were to add some form of ignoring of additional columns, we would have to allocate spare memory to discard the various types of additional data that came back. For example if there was an additional string that came back, and we had nothing we could bind to for it, we'd have to allocate a string and throw it out by pointing to it like that.

The other thing we could _potentially_ do is always pass in the select list where we know it. In the case of `models.Pilots().All()` for example, we know the entire select list we have and could pass that in. But it does make query building more expensive overall since we can no longer use `*`.

There are a few workarounds for this situation:
1. It should be very easy to determine if the schema has changed in the Rails app and automatically roll the Go app when the deployment goes out too. This kind of automated build system is relatively easy to build.
2. Use the `Select()` query mod to only select the data you want. This will allow you to ignore the other fields that may potentially come back by explicitly selecting what your app depends on.

**Comment by Yongke** [comment]
Thanks very much, @aarondl. I understand the situation now.

I also thinked about what you mentioned `The other thing we could potentially do is always pass in the select list where we know it`, this seems not very good idea too, not only because the performance, but we also have situation of deleting of columns (unlikely happen, but exist)

I think solution 2 is better for our situation, however is not so convenient.

**Comment by aarondl** [comment]
Right. So if your database is as fluid as it sounds, it's probably best to do one of the two workarounds I suggested. I do agree that there's a measure of inconvenience here. Ruby's ActiveRecord makes this easier since if you don't consume the field in any way, there's no way for it to break (whether added or deleted) because all the values just go into a `map[string]interface{}` essentially anyway. Not very performant, but definitely is better for this use case.

It's obviously possible to do the same sort of thing even with code generation. But then all fields would have to be methods that would probably have to return a bool like:

```go
p, err := models.Pilots().One()
callSign, has := p.CallSign()
if !has {
  // Do something
}
```

I don't think anyone really wants this API lol.

So with respect I'll close this issue. This is something very interesting because it has implications for HA Go services too. For example you have a database and 3 Go web servers dealing with it. How do you use SQLBoiler and safely add a field to one without experiencing this same problem and/or downtime. Unless you use select statements everywhere (like I suggested) you'll run into this issue where you have to regen and re-deploy in order for something to be compatible with an otherwise compatible database. Maybe sometime down the line we'll have to revisit it though, thanks for bringing it up.

**Comment by Yongke** [comment]
Hi, @aarondl 
I finally choose to provide a patch to workaround with this. It seems work fine and I'm going to deploy it on our product. But not sure it's secure enough, would you please review it for me? Thanks very much.

https://github.com/Yongke/sqlboiler/commit/e8dd2fd801f987c615d5859e465ea4afac06fb92

**Comment by aarondl** [comment]
I'd have made another constant for your special value. That's what sentinel itself is, and straying from that just makes the code more confusing :)

It seems okay to me? Though I'm kinda curious how the `var placeholder interface{}` manages to store any values at all... That seems suspicious. If I were you I would write a test for this to ensure that it's working as expected.

**Comment by aarondl** [comment]
Talked about this a bit on Slack with another gentleman. I decided to re-open this with an investigation into the performance hits that would come from binding into a pooled slice of interface, BEFORE binding to the struct. This would also solve #153. And allow people to be a little bit looser with running sqlboiler code on different levels of migrations.

It also needs some thinking about what happens with other column lists. Inserts etc.

**Comment by aarondl** [comment]
This is solved in v2.6.0
